controllers/operatorpolicy_controller.go

```diff
@@ -1426,187 +1470,346 @@ func (r *OperatorPolicyReconciler) handleInstallPlan(
 		return updateStatus(policy, noInstallPlansCond, noInstallPlansObj(sub.Namespace)), nil
 	}
 
+	var latestInstallPlan *unstructured.Unstructured
+	var latestGeneration int64 = -1
+
+	// Find the latest InstallPlan that references either the installed CSV or the latest CSV available for update.
+	for i := range installPlans {
+		csvs, found, err := unstructured.NestedStringSlice(
+			installPlans[i].Object, "spec", "clusterServiceVersionNames",
+		)
+		if !found || err != nil {
+			continue
+		}
+
+		matchFound := false
+		containsCurrentCSV := false
+
+		for _, csv := range csvs {
+			if csv == sub.Status.CurrentCSV {
+				containsCurrentCSV = true
+				matchFound = true
+
+				break
+			}
+
+			if csv == sub.Status.InstalledCSV {
+				matchFound = true
+
+				break
+			}
+		}
+
+		if !matchFound {
+			continue
+		}
+
+		// If the InstallPlan contains the current CSV and the phase indicates it is installing or complete, always
+		// use that InstallPlan since race conditions in OLM can cause a newer InstallPlan to be created with this CSV.
+		if containsCurrentCSV {
+			phase, _, _ := unstructured.NestedString(installPlans[i].Object, "status", "phase")
+			ipInstalling := string(operatorv1alpha1.InstallPlanPhaseInstalling)
+			ipComplete := string(operatorv1alpha1.InstallPlanPhaseComplete)
```

Comment:
I understand prioritizing the Installing plan, but why also Complete plans? It feels like there can be multiple old InstallPlans that have that one, if the operator is done installing?

Reply:
@JustinKuli if an InstallPlan is complete for `currentCSV` (not `installedCSV`) then that means this InstallPlan successfully installed the latest CSV in the channel for this subscription. Otherwise, due to OLM race conditions, it's possible that there are multiple InstallPlans for `currentCSV` where one is approved and completed and one is awaiting approval.